corgidrp/calibrate_nonlin.py

```diff
@@ -0,0 +1,960 @@
+        
+        # Generate evenly spaced values between 20 and 14000
+        mean_linspace = np.linspace(20, 14000, 1+int((14000-20)/20))
+        
+        # Interpolate/extrapolate the relative gain values
+        interp_func = interp1d(corr_mean_signal_sorted, 
+                        rel_gain_smoothed, kind='linear', fill_value='extrapolate')
+        rel_gain_interp = interp_func(mean_linspace)
+        
+        # Normalize the relative gain to the value at norm_val DN
+        # First, find the index for norm_val DN in mean_linspace
+        idxnorm = np.where(mean_linspace == norm_val)[0][0]
+        normconst = rel_gain_interp[idxnorm]
+        rel_gain_interp /= normconst
+        if (norm_val < temp_min) or (norm_val > temp_max):
+            warnings.warn('norm_val is not between the minimum and maximum values '
+                          'of the means for the current EM gain. Extrapolation '
+                          'will be used for norm_val.')
+        
+        if make_plot:
+            fname = 'non_lin_fit_norm_dn'
+            # Plotting Signal vs. Relative Gain normalized at norm_val DN
+            plt.figure()
+            plt.plot(corr_mean_signal_sorted, rel_gain / normconst, 'o', label='Original Data')
+            plt.ylim([0.95, 1.05])
+            plt.xlim([1, 14000])
+            plt.axhline(1.0, linestyle='--', color='k', linewidth=1)  # horizontal line at 1.0
+            
+            plt.title(f'Signal/fit versus Signal (norm @ {norm_val} DN)')
+            plt.xlabel('Signal (DN)')
+            plt.ylabel('Relative gain')
+            
+            # Plot the interpolated data
+            plt.plot(mean_linspace, rel_gain_interp, 'r-', label='Interpolated Data')
+            plt.legend()
+            plt.savefig(f'{plot_outdir}/{fname}')
+            if verbose:
+                print(f'Figure {fname} stored in {plot_outdir}')
+            if show_plot:
+                plt.show()
+            plt.close()
+        
+        # NOTE: nonlinearity is equal to 1/rel_gain
+        # multiply raw data by 1/rel_gain to correct for nonlinearity
+        temp = 1/rel_gain_interp
+        nonlin.append(temp)
+    
+    # prepare nonlin array
+    nonlin_arr0 = np.transpose(np.array(nonlin))
+    # insert new column at the start of nonlin_arr
+    nonlin_arr1 = np.insert(nonlin_arr0, 0, mean_linspace, axis=1)
+    # select rows that satisfy min/max limits
+    nonlin_arr2 = nonlin_arr1[nonlin_arr1[:, 0] >= min_write]
+    nonlin_arr3 = nonlin_arr2[nonlin_arr2[:, 0] <= max_write]
+    # See data.NonLinearityCalibration doc string for more details:
+    # [0, 1:]: Gain axis values
+    # [1:, 0]: "count" axis value
+    actual_gain_arr = np.insert(actual_gain_arr, 0, np.nan)
+    n_col = len(nonlin_arr3) + 1
+    n_row = len(actual_gain_arr)
+    nonlin_data=np.insert(nonlin_arr3, 0, actual_gain_arr).reshape(n_col,n_row)
+    
+    # Return NonLinearity instance
+    prhd, exthd = create_default_headers()
```

Comment:
Same comment as KGain: we shouldn't fill in dummy headers, we should be using real headers from the input data.

Reply:
Done. Please check for any further suggestions

Reply:
Looks good!